packages/flutter_tools/lib/src/migrate/migrate_compute.dart

```diff
@@ -0,0 +1,625 @@
+  'test', // Files here are typically user owned and flutter-side changes are not relevant.
+  'assets', // Common directory for user assets.
+  'build' // Build artifacts
+];
+
+bool _skipped(String localPath, FileSystem fileSystem, {Set<String>? blacklistPrefixes}) {
+  for (final String path in _skippedFiles) {
+    if (path.replaceAll('/', fileSystem.path.separator) == localPath) {
+      return true;
+    }
+  }
+  if (_skippedFiles.contains(localPath)) {
+    return true;
+  }
+  for (final String dir in _skippedDirectories) {
+    if (localPath.startsWith('${dir.replaceAll('/', fileSystem.path.separator)}${fileSystem.path.separator}')) {
+      return true;
+    }
+  }
+  if (blacklistPrefixes != null) {
+    for (final String prefix in blacklistPrefixes) {
+      if (localPath.startsWith('${prefix}${fileSystem.path.separator}')) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+const List<String> _skippedMergeFileExt = <String>[
+  // Don't merge image files
+  '.png',
+  '.jpg',
+  '.jpeg',
+  '.gif',
+  '.bmp',
+  '.svg',
+  // Don't merge compiled artifacts and executables
+  '.jar',
+  '.so',
+  '.exe',
+];
+
+const Set<String> _alwaysMigrateFiles = <String>{
+  '.metadata', // .metadata tracks key migration information.
+  'android/gradle/wrapper/gradle-wrapper.jar',
+  // Always add .gitignore back in even if user-deleted as it makes it
+  // difficult to migrate in the future and the migrate tool enforces git
+  // usage.
+  '.gitignore',
+};
+
+/// True for files that should not be merged. Typically, images and binary files.
+bool _skippedMerge(String localPath) {
+  for (final String ext in _skippedMergeFileExt) {
+    if (localPath.endsWith(ext) && !_alwaysMigrateFiles.contains(localPath)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Data class holds the common context that is used throughout the steps of a migrate computation.
+class MigrateContext {
+  MigrateContext({
+    required this.migrateResult,
+    required this.flutterProject,
+    required this.blacklistPrefixes,
+    required this.logger,
+    required this.verbose,
+    required this.fileSystem,
+    required this.status,
+    required this.migrateUtils,
+    this.baseProject,
+    this.targetProject,
+  });
+
+  MigrateResult migrateResult;
+  FlutterProject flutterProject;
+  Set<String> blacklistPrefixes;
+  Logger logger;
+  bool verbose;
+  FileSystem fileSystem;
+  Status status;
+  MigrateUtils migrateUtils;
+
+  MigrateBaseFlutterProject? baseProject;
+  MigrateTargetFlutterProject? targetProject;
+}
+
+String getLocalPath(String path, String basePath, FileSystem fileSystem) {
+  return path.replaceFirst(basePath + fileSystem.path.separator, '');
+}
+
+/// Returns a base revision to fallback to in case a true base revision is unknown.
+String _getFallbackBaseRevision(bool allowFallbackBaseRevision, bool verbose, Logger logger, Status status) {
+  if (!allowFallbackBaseRevision) {
+    status.stop();
```

Comment:
You have to be really careful starting and stopping statuses--it will throw if you stop it if it's already stopped, or in the wrong state.  I suggest moving the status starting and stopping to the same scope, in `finally` blocks, like this:
https://github.com/flutter/flutter/blob/db829c1e200823cd655a6e0ca326967ff89d9c4f/packages/flutter_tools/lib/src/cache.dart#L709-L717